File: module5_ws/src/cpp_number_publisher/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int32.hpp"

class NumberCounter: public rclcpp::Node
{
public:
    NumberCounter(): Node("number_counter"), sum(0)
    {
        subscription = this->create_subscription<example_interfaces::msg::Int32>(
            "number",
            10,
            std::bind(&NumberCounter::add_values, this, std::placeholders::_1)
        );
    }

    void add_values(example_interfaces::msg::Int32::SharedPtr msg)
    {
        sum += msg->data;
        RCLCPP_INFO(this->get_logger(), "Received: %d, Sum: %d", msg->data, sum);
    }

private:
    int sum;
    rclcpp::Subscription<example_interfaces::msg::Int32>::SharedPtr subscription;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: module5_ws/src/number_publisher/number_publisher/number_counter.py
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int32

class NumberCounter(Node):
    def __init__(self):
        super().__init__('number_counter')
        self.subscription = self.create_subscription(Int32, 'number', self.add_values, 10)
        self.get_logger().info('Number Counter Node has been started.')
        self.sum_ = 0
    
    def add_values(self, msg):
        self.sum_ += msg.data
        self.get_logger().info(f'Received: {msg.data}, Sum: {self.sum_}')

def main(args=None):
    rclpy.init(args=args)
    number_counter = NumberCounter()
    rclpy.spin(number_counter)
    rclpy.shutdown()

if __name__ == '__main__':
    main()